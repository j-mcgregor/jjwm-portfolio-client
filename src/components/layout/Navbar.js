/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/interactive-supports-focus */
import React, { useState, useEffect } from 'react';
import S from '../styledComponents';
import { dimensions } from '../styledComponents/breakpoints';
import NavList from './NavList';

export const shouldCollapse = (width, breakpoint) => width < breakpoint;

export const shouldSetResponsiveMode = (
  collapsed,
  responsive,
  setMode,
  setOpen
) => {
  if (collapsed && !responsive) setMode(true);

  if (!collapsed && responsive) {
    setOpen(false);
    if (responsive) setMode(false);
  }
};

const Navbar = () => {
  const [windowWidth, setWindowWidth] = useState(window.innerWidth);
  const [responsiveMode, setResponsiveMode] = useState(false);
  const [isOpen, setOpen] = useState(false);
  const overFlow = isOpen ? 'hidden' : 'auto';

  document.body.style.overflowY = overFlow;

  shouldSetResponsiveMode(
    shouldCollapse(windowWidth, dimensions.bsSmall),
    responsiveMode,
    setResponsiveMode,
    setOpen
  );

  const resetModal = () => {
    setOpen(false);
  };

  const resizeWindow = () => {
    setWindowWidth(window.innerWidth);
  };

  useEffect(() => {
    resizeWindow();
    window.addEventListener('resize', resizeWindow);
    return () => window.removeEventListener('resize', resizeWindow);
  }, []);

  return (
    <S.Nav isOpen={isOpen}>
      <nav className="navbar navbar-expand fixed-top flex-md-column flex-row align-items-start">
        <div className="collapse navbar-collapse p-2" id="nav">
          {responsiveMode ? (
            <div className="dropdown">
              <div className="d-flex">
                <i
                  className={`fas fa-${isOpen ? 'times fadeIn' : 'bars'} fa-2x`}
                  role="button"
                  onClick={() => setOpen(!isOpen)}
                />
              </div>
              {isOpen && (
                <div className="fadeIn">
                  <ul>
                    <NavList closeModal={resetModal} />
                  </ul>
                </div>
              )}
            </div>
          ) : (
            <ul className="flex-md-column flex-row navbar-nav w-100 justify-content-sm-around">
              <NavList closeModal={resetModal} />
            </ul>
          )}
        </div>
      </nav>
    </S.Nav>
  );
};

export default Navbar;
