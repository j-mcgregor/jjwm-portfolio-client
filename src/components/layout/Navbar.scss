.navbar {
  padding:0;
  height: 100vh;
  text-align: center;
  position: fixed;

}
.navbar-expand .navbar-collapse {
  width: 16.6666666667%;
}
// 768 -> 720
