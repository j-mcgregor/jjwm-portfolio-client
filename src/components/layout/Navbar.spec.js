import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Navbar, { shouldCollapse, shouldSetResponsiveMode } from './Navbar';

Enzyme.configure({ adapter: new Adapter() });

describe('Navbar', () => {
  let wrapper;
  const props = {
    auth: {
      isAuthenticated: false
    },
    logoutUser: jest.fn()
  };
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);

  beforeEach(() => {
    wrapper = shallow(<Navbar {...props} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Nav helpers', () => {
  describe('shouldCollapse', () => {
    const breakpoint = 575;
    it('should not collapse when screen is bigger than or equal to the predefined breakpoint', () => {
      expect(shouldCollapse(575, breakpoint)).toBe(false);
    });
    it('should collapse when screen is smaller than the predefined breakpoint', () => {
      expect(shouldCollapse(574, breakpoint)).toBe(true);
    });
  });

  describe('shouldSetResponsiveMode', () => {
    let setMode;
    let setOpen;
    beforeEach(() => {
      setMode = jest.fn();
      setOpen = jest.fn();
    });

    it('should return responsive mode = true if collapsed and not yet responsive', () => {
      const collapsed = true;
      const responsive = false;

      shouldSetResponsiveMode(collapsed, responsive, setMode, setOpen);

      expect(setMode).toHaveBeenCalledWith(true);
    });
    it('should return responsive mode = false if not collapsed and responsive', () => {
      const collapsed = false;
      const responsive = true;

      shouldSetResponsiveMode(collapsed, responsive, setMode, setOpen);

      expect(setMode).toHaveBeenCalledWith(false);
    });
    it('should return false from setOpen if not collapsed and responsive', () => {
      const collapsed = false;
      const responsive = true;

      shouldSetResponsiveMode(collapsed, responsive, setMode, setOpen);

      expect(setOpen).toHaveBeenCalledWith(false);
    });
  });
});
