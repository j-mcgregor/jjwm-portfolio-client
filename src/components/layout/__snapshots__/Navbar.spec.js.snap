// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should change responsive state to true if windowWidth < breakpoint 1`] = `ShallowWrapper {}`;

exports[`Navbar should render 1`] = `ShallowWrapper {}`;
